### ISSUE TITLE
Adding metadata to a task templace

### ISSUE THREAD

**Issue by TheCloud** [kanboard/forum#99]
Hello,
I would like to be able to assign each task to a specific customer for billing purposes.
Till now I was using metadata plugin, but it's rather unfriendly.

I was thinking to create a plugin, implementing metadata, in order to be able to track the customer.
I know there is a kit to start a new plugin, but what's after creating the package with the right names ? where to start ? Anyone willing to help ? (I do also have a small budget for that).

Thank you
Julien

**Comment by creecros** [comment]
You want to add an assignment field for customers, or replace existing assignment field for customers?

**Comment by TheCloud** [comment]
Example: http://osticket.com/blog/93

**Comment by creecros** [comment]
Existing assignment field:
![image](https://user-images.githubusercontent.com/26339368/45094848-82514480-b0ea-11e8-9df2-542e1ce82b4c.png)

Sounds to me like you want more than just an Customer Assignment field anyway. You keep saying Metadata, so your best place to start would be with the Metdata Plugin some else already created, and then figuring out what you need to do to make it work the way you want.

**Comment by TheCloud** [comment]
Hi,
I keep on using metadata because for the documentation/product it's the best way (and I agree) to store plugin's data in the same database architecture and modelling in the mind of the author.
I do have already installed the metadata plugin which is very very basic - it's UI is just a form with two "input type text" fields, one for label and one for value that are associated to a task.
The metadata plugin does not implement any changes to the standard task's view template (so you have to open the metadata screen every time)
The metadata plugin does not allow creation of selects, radios, checkboxes or such
The metadata plugin does not allow to tell "add this field to all tasks" - so if you want to have a new field, let's say "gender of a task" (excuse my fantasy), you have to create that metadata this each time for each task.

The goal, basically, is not just to add a "customer" field, but to allow each user to extend the fields that can be set for a task without tweaking the code.

**Comment by creecros** [comment]
no matter what you do, you'll have to edit code.
question is, how much work do you want to put into this?

The metadata plugin does not implement any changes to the standard task's view template (so you have to open the metadata screen every time) - _You will have to create hooks or replace the template on every page you want them to show up_
The metadata plugin does not allow creation of selects, radios, checkboxes or such - _this will complicate things even more, and without having looked at the code for that plugin, my guess is you would have to add a field to designate the type of field each key is, and then build a way to read and present it back to you correctly, as a checkbox or radio._
The metadata plugin does not allow to tell "add this field to all tasks" - so if you want to have a new field, let's say "gender of a task" (excuse my fantasy), you have to create that metadata this each time for each task. - _Well, even if it did, they would be empty values until you added the value of the key to each task, so this is meaningless. You'll have to change that template, so that it no longer relates to the task, but metadata as a whole, and by adding a key, opens up a value input on the templates you want for each task._

You'll also need to create filters for searching...

**Comment by TheCloud** [comment]
- yes, hands on cose should be put by the developer to create the plugin, but no code for the project owner/user to add fields to a task

- more field types make things harder, but we can start simple with a text and extend the plugin on future releases

- adding fields once per project or every time on a task isn’t meaningless - if you have more than one people on a project you can’t tell everybody to add fields with specific names, they will make mistakes and it’s not quick and simple

In terms of templates, let say you create different type of metadata entries:
- type “configuration” that tells the UI to add field A, B and C to project 1,2 and 3
- type “values” that for project 1, for each task, add field A / for project 2 add field B etc

So you can simply make the task view/edit template to cycle adding fields.

Practical example:

NAME.                      VALUE
metadata_field_1_A_name Gender
metadata_field_1_A_Required True
metadata_field_1_A_type Text

This will add a new required  text field for all the tasks of project 1. The field’s name would be “Gender” (but can be translated)

Once you add/edit the task you can tell “Gender” ‘a value like to be “F” or “M”

metadata_field_1_A_value   F

If you need more functionalities, you could add more propeties:
metadata_field_1_A_maxlength 1 
metadata_field_1_A_regexp {1}[F,M]

Etc

Every time you create a new task in project 1, metadata_field_1 is searched and for each “name” found it would add a field...

But for sure you need a UI for the plugin to create items correctly

**Comment by creecros** [comment]
Im working on an issue/project right now, but would be willing to start on this next. Been a few months since I worked on a plugin, so I might be rusty. You can start up repository and add me or I can, up to you.

**Comment by TheCloud** [comment]
The current metata plugin (https://github.com/BlueTeck/kanboard_plugin_metadata) is over (the developer taking care of this went away from the company he was working on and he doesn't use kanboard anymore), we could take it as a start... If you don't mind I would ask him to add us to the current project

Just to be more precise, I checked the database structure and noticed that there are two tables:
project_has_metadata => holding project's metadata pairs => it could hold field's info
task_has_metadata=> holding task's metadata paris => it could hold field's values

**Comment by creecros** [comment]
Either that, or fork it. I took a quick look, noticing the settings metadata, is already in the right direction with field type and required, but not showing up on my test server sidebar with v 1.2.5.

**Comment by creecros** [comment]
Because it's commented out, guess he never finished that part, but it's a good start

**Comment by TheCloud** [comment]
Yes, fine you agree - I will fork and try to implement some ideas, but yet I’m definitively not able to code in that style.. :-)

**Comment by creecros** [comment]
I forked it, and added you as a collaborator: https://github.com/creecros/MetaMagik

Went ahead and corrected all the "new" code daniele lenares added, and got it working.
![image](https://user-images.githubusercontent.com/26339368/45137908-78762280-b178-11e8-939b-a097ac4820cb.png)

it's incomplete in functionality though

**Comment by TheCloud** [comment]
I've checkout (oh sorry, "cloned") the repo into my kanboard - yes I've notices that as kanboard admin I can now add a field (numeric o text) to an entity (project, task, user) and specify also if it's mandatory or not (let's say that the mandatory/optional is not a priority now, do you agree ?)

It seems that it's really close to what I had in mind 🥇

**Comment by creecros** [comment]
It is, but it will take a lot of work to actually incorporate it. It functions, but is not actually connected to the existing features.

I've been brainstorming the original intent, and how to do this, here are my thoughts:

1. Use the Metadata Settings panel to Add Metadata Types 
2. The Metadata Types setup in settings, based on options, will present themselves at the Task Creation/Project Creation/User Creation (and respective Edit Templates)
3. Mandatory/Optional check shouldn't be hard to implement, but not priority
4. Add more types like Boolean to incorporate checkbox, $ to present integers as $
5. Radio Type will not be a priority, as this will prove more difficult with only key/value, maybe add down the road
6. Not sure the original functionality will be needed, i.e. no need for sidebar "Metadata" in the task/project/user any longer, since the Metadata will be edited in the creation and modify templates respectively.

**Comment by creecros** [comment]
We will basically be creating a plugin for "Custom Fields"

**Comment by TheCloud** [comment]
1. I agree, it make sense for "user" and "project" entity, but for the "task" entity it should be allocated to project's owner
2. it ok for me to have them in the creation/edit
3. Agree it's not priority
4. I still think "select" (ie list of options from which to choose) is mor interesting than integer
5. I agree it's not priority
6. I agree, if the added field is appended to the original template for user/project/task it's not needed in the sidebar

**Comment by creecros** [comment]
1.) actually, this is proving challenging in my mind, because task level is not project/task level, it is all tasks. it would be simple enough to give project owners the ability to edit the types for tasks, but it would not be project dependent. The entity can't be tasks within a certain project. You could add another subentity field to metadata_type table, but then i think your going down a rabbit hole of checks to perform in a boat load of controllers. maybe there is a simple solution, but not currently coming to mind.

**Comment by TheCloud** [comment]
1. yep
2. I agree, it's more complex, can definitively go "later", I mean it's not needed right now. Having custom fields it's nice anyway

**Comment by TheCloud** [comment]
I tested the new plugin MetaMagik and works charmy :-)

**Comment by creecros** [comment]
It's getting there, lots more to do though

**Comment by TheCloud** [comment]
It’s there!
